examples/slice/slice_utils.py

```diff
@@ -0,0 +1,341 @@
+import firedrake as fd
+import numpy as np
+from firedrake import op2
+
+
+def maximum(f):
+    fmax = op2.Global(1, [-1e50], dtype=float)
+    op2.par_loop(op2.Kernel("""
+static void maxify(double *a, double *b) {
+    a[0] = a[0] < b[0] ? b[0] : a[0];
+}
+""", "maxify"), f.dof_dset.set, fmax(op2.MAX), f.dat(op2.READ))
+    return fmax.data[0]
+
+
+def minimum(f):
+    fmin = op2.Global(1, [1e50], dtype=float)
+    op2.par_loop(op2.Kernel("""
+static void minify(double *a, double *b) {
+    a[0] = a[0] > b[0] ? b[0] : a[0];
+}
+""", "minify"), f.dof_dset.set, fmin(op2.MIN), f.dat(op2.READ))
+    return fmin.data[0]
+
+
+def pi_formula(rho, theta, R_d, p_0, kappa):
+    return (rho * R_d * theta / p_0) ** (kappa / (1 - kappa))
+
+
+def rho_formula(pi, theta, R_d, p_0, kappa):
+    return p_0*pi**((1-kappa)/kappa)/R_d/theta
+
+
+def hydrostatic_rho(Vv, V2, mesh, thetan, rhon, pi_boundary,
+                    cp, R_d, p_0, kappa, g, Up,
+                    top=False, Pi=None):
+    # Calculate hydrostatic Pi, rho
+    W_h = Vv * V2
+    wh = fd.Function(W_h)
+    n = fd.FacetNormal(mesh)
+    dv, drho = fd.TestFunctions(W_h)
+
+    v, Pi0 = fd.TrialFunctions(W_h)
+
+    Pieqn = (
+        (cp*fd.inner(v, dv) - cp*fd.div(dv*thetan)*Pi0)*fd.dx
+        + drho*fd.div(thetan*v)*fd.dx
+    )
+
+    if top:
+        bmeasure = fd.ds_t
+        bstring = "bottom"
+    else:
+        bmeasure = fd.ds_b
+        bstring = "top"
+
+    zeros = []
+    for i in range(Up.ufl_shape[0]):
+        zeros.append(fd.Constant(0.))
+
+    L = -cp*fd.inner(dv, n)*thetan*pi_boundary*bmeasure
+    L -= g*fd.inner(dv, Up)*fd.dx
+    bcs = [fd.DirichletBC(W_h.sub(0), zeros, bstring)]
+
+    PiProblem = fd.LinearVariationalProblem(Pieqn, L, wh, bcs=bcs)
+
+    lu_params = {
+        'snes_monitor': None,
+        'snes_stol': 1.0e-50,
+        'ksp_monitor': None,
+        'snes_converged_reason': None,
+        'mat_type': 'aij',
+        'pc_type': 'lu',
+        "pc_factor_mat_ordering_type": "rcm",
+    }
```

Comment:
Does this solver need to have the monitor and converged reasons switched on? It gets messy when every time rank is trying to print at once. If that output is still useful, then maybe only one time-rank should print, or pass a `verbose` flag to the function.